src/main/java/com/bina/lrsim/simulator/Simulator.java

```diff
@@ -254,21 +274,28 @@ public int simulateWithNoisyAdapter(final String path, final String movieName, f
          *
          *
          */
-        // in some sequence drawer, such as fragment mode, the sequence can be much longer than read length, in this case we set the passes to fragment length
-        if(multiPassSpec.fragmentLength < Heuristics.READLENGTH_RESCUE_FRACTION * sampledReferenceSequence.length) {
-          //why starts from 1 instead of 0?
-          for (int ii = 1; ii + 1 < insertLengths.length; ++ii) {
-            if (insertLengths[ii] < sampledReferenceSequence.length) {
-              insertLengths[ii] = sampledReferenceSequence.length;
+        // correct insert lengths if the drawn fragment is shorter, the fractional change might not be realistic, but it avoids crazy coverage in fragment mode
+        // if we want to estimate number of passes by insert lengths, we break up the lengths into chunks
+        // that are equal to lengths of template sequence drawn
+        if (spec.isEstimateNPByLength() && spec.getSequencingMode() == SequencingMode.fragment) {
+          //right now we assume the sampled lengths contain adapter lengths
+          //i.e. each multiples of adapter + template except for the first
+          //one
+          insertLengths = breakUpLengths(insertLengths[0], sampledReferenceSequence.length, Heuristics.SMRT_ADAPTOR_STRING.length);
```

Comment:
The `insertLengths` are summed up here: https://github.com/bioinform/longislnd/pull/31/files#diff-ef9f42154f4ec285ec5b7867845d9ae0R78, right after they are loaded.